File: src/main/java/net/ccbluex/liquidbounce/LiquidBounce.java
package net.ccbluex.liquidbounce;

import com.google.gson.JsonObject;
import com.google.gson.JsonParser;
import me.aquavit.liquidsense.LiquidSense;
import me.aquavit.liquidsense.utils.client.ClientUtils;
import me.aquavit.liquidsense.utils.forge.BlocksTab;
import me.aquavit.liquidsense.utils.forge.ExploitsTab;
import me.aquavit.liquidsense.utils.forge.HeadsTab;
import me.aquavit.liquidsense.utils.login.UserUtils;
import me.aquavit.liquidsense.utils.mc.ClassUtils;
import me.aquavit.liquidsense.event.events.ClientShutdownEvent;
import me.aquavit.liquidsense.event.EventManager;
import me.aquavit.liquidsense.command.CommandManager;
import net.ccbluex.liquidbounce.features.module.ModuleManager;
import me.aquavit.liquidsense.ui.client.clickgui.neverlose.Main;
import me.aquavit.liquidsense.ui.client.gui.elements.AntiForge;
import me.aquavit.liquidsense.ui.client.gui.elements.BungeeCordSpoof;
import me.aquavit.liquidsense.file.FileManager;
import me.aquavit.liquidsense.script.ScriptManager;
import me.aquavit.liquidsense.script.remapper.Remapper;
import me.aquavit.liquidsense.ui.client.gui.GuiAltManager;
import me.aquavit.liquidsense.ui.client.hud.HUD;
import me.aquavit.liquidsense.ui.font.Fonts;
import me.aquavit.liquidsense.utils.client.InventoryUtils;
import me.aquavit.liquidsense.utils.client.RotationUtils;
import me.aquavit.liquidsense.utils.misc.HttpUtils;
import net.minecraft.util.ResourceLocation;
import net.minecraftforge.fml.relauncher.Side;
import net.minecraftforge.fml.relauncher.SideOnly;

@SideOnly(Side.CLIENT)
public class LiquidBounce {

    public static LiquidSense liquidSense;

    public static final LiquidBounce INSTANCE = new LiquidBounce();

    public static final String CLIENT_NAME = "LiquidSense";
    public static final String CLIENT_VERSION = "Dev";
    public static final String CLIENT_CREATOR = "CCBlueX";
    public static final String MINECRAFT_VERSION = "1.8.9";
    public static final String CLIENT_CLOUD = "https://cloud.liquidbounce.net/LiquidBounce";
    public static final String CLIENT_RESOURCE = "https://cdn.jsdelivr.net/gh/BlogResourceRepositories/ClientResource@main/";
    public static final Boolean IN_DEV = true;

    public boolean isStarting;

    public static ModuleManager moduleManager;
    public static CommandManager commandManager;
    public static EventManager eventManager;
    public static FileManager fileManager;
    public static ScriptManager scriptManager;

    public static HUD hud;
    public static Main neverlose;

    public int latestVersion;

    public ResourceLocation background;

    public void startClient() {
        this.isStarting = true;

        ClientUtils.getLogger().info("Starting "+CLIENT_NAME+" "+CLIENT_VERSION+", by "+CLIENT_CREATOR);

        LiquidSense liquidSense = LiquidBounce.liquidSense = new LiquidSense(this);
        liquidSense.onStarting();

        fileManager = new FileManager();

        EventManager eventManager = LiquidBounce.eventManager = new EventManager();

        eventManager.registerListener(new RotationUtils());
        eventManager.registerListener(new AntiForge());
        eventManager.registerListener(new BungeeCordSpoof());
        eventManager.registerListener(new InventoryUtils());

        commandManager = new CommandManager();

        Fonts.loadFonts();

        moduleManager = new ModuleManager();
        moduleManager.registerModules();

        try {
            Remapper.INSTANCE.loadSrg();
            scriptManager = new ScriptManager();
            scriptManager.loadScripts();
            scriptManager.enableScripts();
        } catch (Throwable throwable) {
            ClientUtils.getLogger().error("Failed to load scripts.", throwable);
        }

        commandManager.registerCommands();

        fileManager.loadConfigs(fileManager.modulesConfig, fileManager.valuesConfig, fileManager.accountsConfig,
                fileManager.friendsConfig, fileManager.xrayConfig, fileManager.shortcutsConfig);

        neverlose = new Main();

        if (ClassUtils.hasForge()) {
            new BlocksTab();
            new ExploitsTab();
            new HeadsTab();
        }

        hud = HUD.Companion.createDefault();
        fileManager.loadConfig(fileManager.hudConfig);

        ClientUtils.disableFastRender();

        try {
            // Read versions json from cloud
            JsonObject jsonObj = new JsonParser().parse(HttpUtils.get(CLIENT_CLOUD + "/versions.json")).getAsJsonObject();
            // Check json is valid object and has current minecraft version
            if (jsonObj.isJsonObject() && jsonObj.has(MINECRAFT_VERSION)) {
                // Get official latest client version
                latestVersion = jsonObj.get(MINECRAFT_VERSION).getAsInt();
            }
        } catch (Throwable exception) {
            ClientUtils.getLogger().error("Failed to check for updates.");
        }

        LoadAltManagerSkin();

        GuiAltManager.loadGenerators();

        fileManager.loadConfigs(fileManager.setnameConfig);

        liquidSense.onStarted();

        isStarting = false;
    }

    public final void stopClient() {
        LiquidBounce.eventManager.callEvent(new ClientShutdownEvent());
        LiquidBounce.fileManager.saveAllConfigs();
    }

    public final void LoadAltManagerSkin(){

        ClientUtils.getLogger().info("Loading AltManagerSkin.");

        for (int id = 0; id < LiquidBounce.fileManager.accountsConfig.altManagerMinecraftAccounts.size(); ++id) {
            if (!LiquidBounce.fileManager.accountsConfig.altManagerMinecraftAccounts.get(id).isCracked() && !GuiAltManager.skin.containsKey(id))
                GuiAltManager.skin.put(id, UserUtils.getPlayerSkin(UserUtils.getUUID(LiquidBounce.fileManager.accountsConfig.altManagerMinecraftAccounts.get(id).getAccountName())));
        }
    }

    public final LiquidSense getLiquidSense() {
        return LiquidBounce.liquidSense;
    }

    public final void setLiquidSense(LiquidSense liquidSense) {
        LiquidBounce.liquidSense = liquidSense;
    }

    public final boolean isStarting() {
        return isStarting;
    }

    public final void setStarting(boolean state) {
        isStarting = state;
    }

    public final ModuleManager getModuleManager() {
        return LiquidBounce.moduleManager;
    }

    public final void setModuleManager(ModuleManager moduleManager) {
        LiquidBounce.moduleManager = moduleManager;
    }

    public final CommandManager getCommandManager() {
        return LiquidBounce.commandManager;
    }

    public final void setCommandManager(CommandManager commandManager) {
        LiquidBounce.commandManager = commandManager;
    }

    public final EventManager getEventManager() {
        return LiquidBounce.eventManager;
    }

    public final void setEventManager(EventManager eventManager) {
        LiquidBounce.eventManager = eventManager;
    }

    public final FileManager getFileManager() {
        return LiquidBounce.fileManager;
    }

    public final void setFileManager(FileManager fileManager) {
        LiquidBounce.fileManager = fileManager;
    }

    public final ScriptManager getScriptManager() {
        return LiquidBounce.scriptManager;
    }

    public final void setScriptManager(ScriptManager scriptManager) {
        LiquidBounce.scriptManager = scriptManager;
    }

    public final HUD getHud() {
        return hud;
    }

    public final void setHud(HUD hud) {
        hud = hud;
    }

    public final Main getNeverlose() {
        return neverlose;
    }

    public final void setNeverlose(Main main) {
        neverlose = main;
    }

    public final int getLatestVersion() {
        return latestVersion;
    }

    public final void setLatestVersion(int version) {
        latestVersion = version;
    }

    public final ResourceLocation getBackground() {
        return background;
    }

    public final void setBackground(ResourceLocation resourceLocation) {
        background = resourceLocation;
    }
}

File: src/main/java/net/ccbluex/liquidbounce/LiquidBounce.kt
/*
 * LiquidBounce Hacked Client
 * A free open source mixin-based injection hacked client for Minecraft using Minecraft Forge.
 * https://github.com/CCBlueX/LiquidBounce/
 */
package net.ccbluex.liquidbounce

import CommandManager
import com.google.gson.JsonObject
import com.google.gson.JsonParser
import me.aquavit.liquidsense.LiquidSense
import me.aquavit.liquidsense.utils.forge.BlocksTab
import me.aquavit.liquidsense.utils.forge.ExploitsTab
import me.aquavit.liquidsense.utils.forge.HeadsTab
import net.ccbluex.liquidbounce.event.ClientShutdownEvent
import net.ccbluex.liquidbounce.event.EventManager
import net.ccbluex.liquidbounce.features.module.ModuleManager
import net.ccbluex.liquidbounce.features.special.AntiForge
import net.ccbluex.liquidbounce.features.special.BungeeCordSpoof
import net.ccbluex.liquidbounce.file.FileManager
import net.ccbluex.liquidbounce.script.ScriptManager
import net.ccbluex.liquidbounce.script.remapper.Remapper.loadSrg

import net.ccbluex.liquidbounce.ui.client.gui.GuiAltManager
import net.ccbluex.liquidbounce.ui.client.clickgui.ClickGui
import net.ccbluex.liquidbounce.ui.client.hud.HUD
import net.ccbluex.liquidbounce.ui.client.hud.HUD.Companion.createDefault
import net.ccbluex.liquidbounce.ui.font.Fonts
import me.aquavit.liquidsense.utils.mc.ClassUtils
import net.ccbluex.liquidbounce.ui.client.miscible.Miscible
import net.ccbluex.liquidbounce.ui.client.neverlose.Main
import net.ccbluex.liquidbounce.utils.ClientUtils
import net.ccbluex.liquidbounce.utils.InventoryUtils
import net.ccbluex.liquidbounce.utils.RotationUtils
import net.ccbluex.liquidbounce.utils.misc.HttpUtils
import net.minecraft.util.ResourceLocation

object LiquidBounce {

    lateinit var liquidSense: LiquidSense;

    // Client information
    const val CLIENT_NAME = "LiquidSense"
    const val CLIENT_VERSION = "Dev"
    const val CLIENT_CREATOR = "CCBlueX"
    const val MINECRAFT_VERSION = "1.8.9"
    const val CLIENT_CLOUD = "https://cloud.liquidbounce.net/LiquidBounce"
    const val CLIENT_RESOURCE = "https://cdn.jsdelivr.net/gh/BlogResourceRepositories/ClientResource@main/"



    var isStarting = false

    // Managers
    lateinit var moduleManager: ModuleManager
    lateinit var commandManager: CommandManager
    lateinit var eventManager: EventManager
    lateinit var fileManager: FileManager
    lateinit var scriptManager: ScriptManager

    // HUD & ClickGUI
    lateinit var hud: HUD

    lateinit var clickGui: ClickGui
    lateinit var miscible: Miscible
    lateinit var neverlose : Main

    // Update information
    var latestVersion = 0

    // Menu Background
    var background: ResourceLocation? = null

    /**
     * Execute if client will be started
     */
    fun startClient() {
        isStarting = true

        ClientUtils.getLogger().info("Starting $CLIENT_NAME $CLIENT_VERSION, by $CLIENT_CREATOR")

        liquidSense = LiquidSense(this)
        liquidSense.onStarting()

        // Create file manager
        fileManager = FileManager()

        // Crate event manager
        eventManager = EventManager()

        // Register listeners
        eventManager.registerListener(RotationUtils())
        eventManager.registerListener(AntiForge())
        eventManager.registerListener(BungeeCordSpoof())
        eventManager.registerListener(InventoryUtils())

        // Create command manager
        commandManager = CommandManager()

        // Load client fonts
        Fonts.loadFonts()

        // Setup module manager and register modules
        moduleManager = ModuleManager()
        moduleManager.registerModules()

        // Remapper
        try {
            loadSrg()

            // ScriptManager
            scriptManager = ScriptManager()
            scriptManager.loadScripts()
            scriptManager.enableScripts()
        } catch (throwable: Throwable) {
            ClientUtils.getLogger().error("Failed to load scripts.", throwable)
        }

        // Register commands
        commandManager.registerCommands()

        // Load configs
        fileManager.loadConfigs(fileManager.modulesConfig, fileManager.valuesConfig, fileManager.accountsConfig,
                fileManager.friendsConfig, fileManager.xrayConfig, fileManager.shortcutsConfig)

        // ClickGUI
        clickGui = ClickGui()
        miscible = Miscible()
        neverlose = Main()
        fileManager.loadConfig(fileManager.clickGuiConfig)

        // Tabs (Only for Forge!)
        if (ClassUtils.hasForge()) {
            BlocksTab()
            ExploitsTab()
            HeadsTab()
        }

        // Set HUD
        hud = createDefault()
        fileManager.loadConfig(fileManager.hudConfig)

        // Disable optifine fastrender
        ClientUtils.disableFastRender()

        try {
            // Read versions json from cloud
            val jsonObj = JsonParser()
                    .parse(HttpUtils.get("$CLIENT_CLOUD/versions.json"))

            // Check json is valid object and has current minecraft version
            if (jsonObj is JsonObject && jsonObj.has(MINECRAFT_VERSION)) {
                // Get offical latest client version
                latestVersion = jsonObj[MINECRAFT_VERSION].asInt
            }
        } catch (exception: Throwable) { // Print throwable to console
            ClientUtils.getLogger().error("Failed to check for updates.", exception)
        }

        // Load generators
        GuiAltManager.loadGenerators()

        fileManager.loadConfigs(fileManager.setnameConfig)


        liquidSense.onStarted()

        // Set is starting status
        isStarting = false
    }

    /**
     * Execute if client will be stopped
     */
    fun stopClient() {

        // Call client shutdown
        eventManager.callEvent(ClientShutdownEvent())

        // Save all available configs
        fileManager.saveAllConfigs()

        // Shutdown discord rpc
        //clientRichPresence.shutdown()
    }

}